pkg/receive/hashring.go

```diff
@@ -66,49 +66,106 @@ func hash(tenant string, ts *prompb.TimeSeries) uint64 {
 	return xxhash.Sum64(b)
 }
 
-// simpleHashring represents a group of hosts handling write requests.
-type simpleHashring struct {
-	targetgroup.Group
+// SingleNodeHashring always returns the same node.
+type SingleNodeHashring string
+
+// Get implements the Hashring interface.
+func (s SingleNodeHashring) Get(_ string, _ *prompb.TimeSeries) (string, error) {
+	return string(s), nil
 }
 
-// GetHost returns a hostname to handle the given tenant and time series.
-func (s *simpleHashring) GetHost(tenant string, ts *prompb.TimeSeries) (string, error) {
+// simpleHashring represents a group of nodes handling write requests.
+type simpleHashring []string
+
+// Get returns a target to handle the given tenant and time series.
+func (s simpleHashring) Get(tenant string, ts *prompb.TimeSeries) (string, error) {
 	// Always return nil here to implement the Hashring interface.
-	return string(s.Targets[hash(tenant, ts)%uint64(len(s.Targets))][model.AddressLabel]), nil
+	return s[hash(tenant, ts)%uint64(len(s))], nil
 }
 
 // matchingHashring represents a set of hashrings.
 // Which hashring to use is determined by the matcher.
 type matchingHashring struct {
 	cache     map[string]Hashring
-	hashrings map[string]Hashring
+	hashrings []Hashring
 	matcher   Matcher
+	cfg       []HashringConfig
+
+	// We need a mutex to guard concurrent access
+	// to the cache map, as this is both written to
+	// and read from.
+	mu sync.Mutex
```

Comment:
we very rarely write to the cache and often read from it, this should be a RWMutex

Reply:
yes absolutely :+1: